File: solutions/SingleThreadedServer/app/src/main/java/ro/pub/cs/systems/eim/lab06/singlethreadedserver/network/ServerThread.java
package ro.pub.cs.systems.eim.lab06.singlethreadedserver.network;

import android.util.Log;
import android.widget.EditText;

import java.io.IOException;
import java.io.PrintWriter;
import java.net.ServerSocket;
import java.net.Socket;

import ro.pub.cs.systems.eim.lab06.singlethreadedserver.general.Constants;
import ro.pub.cs.systems.eim.lab06.singlethreadedserver.general.Utilities;

public class ServerThread extends Thread {

    private boolean isRunning;

    private ServerSocket serverSocket;
    private final EditText serverTextEditText;

    public ServerThread(EditText serverEditText) {
        this.serverTextEditText = serverEditText;
    }

    public void startServer() {
        isRunning = true;
        start();
        Log.v(Constants.TAG, "startServer() method invoked " + serverSocket);
    }

    public void stopServer() {
        isRunning = false;
        try {
            if (serverSocket != null) {
                serverSocket.close();
            }
        } catch (IOException ioException) {
            Log.e(Constants.TAG, "An exception has occurred: " + ioException.getMessage());
        }
        Log.v(Constants.TAG, "stopServer() method invoked");
    }

    @Override
    public void run() {
        try {
            serverSocket = new ServerSocket(Constants.SERVER_PORT);
            while (isRunning) {
                Socket socket = serverSocket.accept();
                Log.v(Constants.TAG, "Connection opened with " + socket.getInetAddress() + ":" + socket.getLocalPort());
                try {
                    Thread.sleep(3000);
                } catch (InterruptedException interruptedException) {
                    Log.e(Constants.TAG, "An exception has occurred: " + interruptedException.getMessage());
                }
                PrintWriter printWriter = Utilities.getWriter(socket);
                printWriter.println(serverTextEditText.getText().toString());
                socket.close();
                Log.v(Constants.TAG, "Connection closed");
            }
        } catch (IOException ioException) {
            Log.e(Constants.TAG, "An exception has occurred: " + ioException.getMessage());
        }
    }
}

File: solutions-kotlin/SingleThreadedServer/app/src/main/java/ro/pub/cs/systems/eim/lab06/singlethreadedserver/network/ServerThread.kt
package ro.pub.cs.systems.eim.lab06.singlethreadedserver.network

import android.util.Log
import android.widget.EditText
import ro.pub.cs.systems.eim.lab06.singlethreadedserver.general.Constants
import ro.pub.cs.systems.eim.lab06.singlethreadedserver.general.Utilities.getWriter
import java.io.IOException
import java.net.ServerSocket

class ServerThread(private val serverTextEditText: EditText) : Thread() {
    private var isRunning = false

    private var serverSocket: ServerSocket? = null

    fun startServer() {
        isRunning = true
        start()
        Log.v(Constants.TAG, "startServer() method invoked " + serverSocket)
    }

    fun stopServer() {
        isRunning = false
        try {
            if (serverSocket != null) {
                serverSocket!!.close()
            }
        } catch (ioException: IOException) {
            Log.e(Constants.TAG, "An exception has occurred: " + ioException.message)
        }
        Log.v(Constants.TAG, "stopServer() method invoked")
    }

    override fun run() {
        try {
            serverSocket = ServerSocket(Constants.SERVER_PORT)
            while (isRunning) {
                val socket = serverSocket!!.accept()
                Log.v(
                    Constants.TAG,
                    "Connection opened with " + socket.getInetAddress() + ":" + socket.getLocalPort()
                )
                try {
                    sleep(3000)
                } catch (interruptedException: InterruptedException) {
                    Log.e(
                        Constants.TAG,
                        "An exception has occurred: " + interruptedException.message
                    )
                }
                val printWriter = getWriter(socket)
                printWriter.println(serverTextEditText.getText().toString())
                socket.close()
                Log.v(Constants.TAG, "Connection closed")
            }
        } catch (ioException: IOException) {
            Log.e(Constants.TAG, "An exception has occurred: " + ioException.message)
        }
    }
}
